### DISCUSSION TITLE
Fill up white inside shape, place object inside same shape

### DISCUSSION THREAD

**Discussion by TNNSLAB** [ImageMagick/ImageMagick#6810] [Help] [answered]
### ImageMagick version

7.1.0-4 Q16 x86_64

### Operating system, version and so on

Mac OS 10.14.6

### Description

I have a lot of rackets where stencils where placed using photoshop and they are all over the place...

<img width="1047" alt="inconsistent" src="https://github.com/ImageMagick/ImageMagick/assets/46636486/6454755e-06be-43b7-8a8d-fc9bcac77ce3">

So I need to fill up the stencil part of the racket with white color 255.

Than place the stencils back to the racket and position - center

![wilson](https://github.com/ImageMagick/ImageMagick/assets/46636486/26e8f9d7-f69a-40a8-b465-3ad49356229a)

Here is the stencil.
![stencil](https://github.com/ImageMagick/ImageMagick/assets/46636486/e29ed8ae-5811-404a-beb7-c62ed8d68294)

Any help is appreciated. Speaking about a lot of images...

**Comment by snibgo** [comment]
> I have a lot of rackets ...

Does the stencil need to be in the same place for each racket? If so, then the command could be something like:
```
magick racket.png stencil.png -geometry +200+300 out.png
```
The geometry coordinates are probably wrong. You don't show a sample input racket.

**Reply by TNNSLAB** [reply]
@snibgo how do i fill the existing stencil with white ?  the stencil needs to be almost the same place all the time...

**Reply by TNNSLAB** [reply]
sample input racket is there...wilson.jpg

**Comment by snibgo** [comment]
All your images you have shown already have stencils. 

> how do i fill the existing stencil with white ?

What needs to be filled with white? Perhaps the input rackets, which you haven't shown, have transparency. A JPG image has no transparency.

**Reply by TNNSLAB** [reply]
Correct. All images have white. They were edited in Photoshop. 

My mission is to go through a lot of JPG's and try to fix them. PSD's are nowhere to find. 

I was thinking I can fill up the area of the stencil with white and than place the stencil.png somewhere around 330 200 or so. 

This way all my rackets would have a more consistent stencil. 

I can composite the stencil.png into that desired position. But I don't know how to fill the existing bad positioned stencil with white.

**Comment by fmw42** [comment]
I don't understand why you are trying to do that. If your rackets already have white with the red strings, why are you trying to covert the strings with white and then put the same shaped strings back?

**Reply by TNNSLAB** [reply]
Because if you check the very first image with 4 rackets, you can see how inconsistent is that stencil. Size of the stencil and it's position is all over...

**Comment by fmw42** [comment]
Are the racket shapes/sizes all the same -- is the oval always the same size and position on your image?  If so, just composite white over that area.  Make a mask of the oval area as white on transparent background and overlay that

**Reply by TNNSLAB** [reply]
There's a bit of fluctuation but it's minimal...

Would you mind showing an example of your idea ? Thank you.

**Reply by fmw42** [reply, marked as answer]
OK, what I did was to draw a green oval (using Photoshop) inside your racket of slightly smaller shape than your racket so that if you have some slight variations in racket size or position, it won't matter.  Redraw smaller or larger as desired. You should be able to use GIMP or Photoshop to draw the oval.

![racket_green](https://github.com/ImageMagick/ImageMagick/assets/13458218/6a0ab6b1-3be8-4b02-ba32-ffdf9310ce24)

I then made a white mask with transparent background. (You will have to download to see the transparency as it all will look white here).

`magick racket_green.png -color-threshold "rgb(0,250,0)-rgb(0,255,0)" -alpha shape y.png`

![y](https://github.com/ImageMagick/ImageMagick/assets/13458218/3dd83363-14e1-4de1-ad0f-0fc1d320ea75)

I then just composited it over your racket.

`magick racket.jpg y.png -compose over -composite racket_fixed.png`

![racket_fixed](https://github.com/ImageMagick/ImageMagick/assets/13458218/7e8a4f35-d99b-44a2-b2d1-5f692f0ae59a)

**Reply by TNNSLAB** [reply]
@fmw42 Awesome ! one last thing...now that we know the size of the oval shape...can i place the stencil in the middle of it ? right now i'm doing the composite like this : 

```composite -geometry +330+2000 stencil.png racket.jpg output.jpg```

**Reply by fmw42** [reply]
What do you mean "in the middle"?  In the middle of the racket **automatically**?  Perhaps by making the mask smaller, crop about the racket oval plus a little. Then use compare to do subimage-search to find the best location. I am not sure this will work, but you can try.  Perhaps make a second mask that has the string pattern in it that is the same size as the cropped white oval. Use the second mask with compare so that it matches to the string pattern. Then use those coordinates to overlay the white oval mask.
